File: apps/api/src/modules/users/application/users.service.ts
import { Injectable } from '@nestjs/common';
import { Algorithm, hash, verify } from '@node-rs/argon2';
import { randomBytes } from 'node:crypto';
import type {
  AccountExport,
  AuthenticatedUser,
  CursorPaginatedResponse,
  ReviewDto,
  UpdateProfileRequest,
  UserProfile,
} from '@gamescore/types';

import { clampLimit, cursorMeta, DEFAULT_CURSOR_SIZE } from '../../../common/http/pagination';
import { BadRequestError, NotFoundError, UnauthorizedError } from '../../../common/errors/app.exception';
import { ERROR_CODES } from '../../../common/errors/error-codes';
import { PrismaService } from '../../../common/prisma/prisma.service';
import { toAuthenticatedUser, toUserProfile } from '../mappers/user.mapper';
import { UserRepository } from '../repositories/user.repository';
import { ReviewRepository } from '../../reviews/repositories/review.repository';
import { toReviewDto } from '../../reviews/mappers/review.mapper';
import type { AuthUser } from '../../auth/domain/auth-user';

const ARGON = {
  algorithm: Algorithm.Argon2id,
  memoryCost: 19_456,
  timeCost: 2,
  parallelism: 1,
} as const;

@Injectable()
export class UsersService {
  constructor(
    private readonly users: UserRepository,
    private readonly reviews: ReviewRepository,
    private readonly prisma: PrismaService,
  ) {}

  async getProfile(username: string): Promise<UserProfile> {
    const user = await this.users.findProfileByUsername(username);
    if (!user) {
      throw new NotFoundError(ERROR_CODES.USER_NOT_FOUND, 'User not found');
    }

    const [reviewAgg, usefulVotes, hours] = await Promise.all([
      this.prisma.review.groupBy({
        by: ['recommended'],
        where: { userId: user.id, deletedAt: null, status: 'PUBLISHED' },
        _count: { _all: true },
      }),
      this.prisma.review.aggregate({
        where: { userId: user.id, deletedAt: null },
        _sum: { usefulCount: true },
      }),
      this.prisma.review.aggregate({
        where: { userId: user.id, deletedAt: null, hoursPlayed: { not: null } },
        _sum: { hoursPlayed: true },
      }),
    ]);

    let recommended = 0;
    let notRecommended = 0;
    for (const row of reviewAgg) {
      if (row.recommended) recommended += row._count._all;
      else notRecommended += row._count._all;
    }
    const totalReviews = recommended + notRecommended;

    return toUserProfile(user, {
      totalReviews,
      recommendedCount: recommended,
      notRecommendedCount: notRecommended,
      recommendationPercentage:
        totalReviews === 0 ? 0 : Math.round((recommended / totalReviews) * 10000) / 100,
      usefulVotesReceived: usefulVotes._sum.usefulCount ?? 0,
      gamesReviewed: totalReviews,
      totalHoursPlayed: hours._sum.hoursPlayed ?? 0,
    });
  }

  async listReviews(
    username: string,
    cursor: string | undefined,
    limit: number | undefined,
    viewer?: AuthUser | null,
  ): Promise<CursorPaginatedResponse<ReviewDto>> {
    const user = await this.users.findProfileByUsername(username);
    if (!user) {
      throw new NotFoundError(ERROR_CODES.USER_NOT_FOUND, 'User not found');
    }

    const pageSize = clampLimit(limit, DEFAULT_CURSOR_SIZE);
    const { items, nextCursor } = await this.reviews.listByUser(user.id, cursor, pageSize);
    const context = viewer
      ? await this.reviews.viewerContext(
          items.map((item) => item.id),
          viewer.id,
        )
      : { votes: new Map<string, boolean>(), reports: new Set<string>() };

    return {
      items: items.map((item) =>
        toReviewDto(item, {
          viewerId: viewer?.id,
          viewerVote: context.votes.has(item.id) ? context.votes.get(item.id)! : null,
          viewerHasReported: context.reports.has(item.id),
        }),
      ),
      meta: cursorMeta(pageSize, nextCursor),
    };
  }

  async updateMe(userId: string, input: UpdateProfileRequest): Promise<AuthenticatedUser> {
    const displayName =
      input.displayName === undefined ? undefined : emptyToNull(input.displayName);
    const bio = input.bio === undefined ? undefined : emptyToNull(input.bio);
    const avatarUrl =
      input.avatarUrl === undefined ? undefined : this.normaliseAvatarUrl(input.avatarUrl);

    const updated = await this.users.updateProfile(userId, { displayName, bio, avatarUrl });
    return toAuthenticatedUser(updated);
  }

  async exportMe(userId: string): Promise<AccountExport> {
    const user = await this.prisma.user.findFirst({
      where: { id: userId, deletedAt: null },
    });
    if (!user) {
      throw new NotFoundError(ERROR_CODES.USER_NOT_FOUND, 'Account not found');
    }

    const [reviews, votes, reports, reputationEvents] = await Promise.all([
      this.prisma.review.findMany({
        where: { userId },
        include: { game: { select: { slug: true, name: true } } },
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.reviewVote.findMany({
        where: { userId },
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.reviewReport.findMany({
        where: { reporterId: userId },
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.reputationEvent.findMany({
        where: { userId },
        orderBy: { createdAt: 'desc' },
      }),
    ]);

    return {
      exportedAt: new Date().toISOString(),
      account: {
        email: user.email,
        username: user.username,
        displayName: user.displayName,
        bio: user.bio,
        avatarUrl: user.avatarUrl,
        reputationScore: user.reputationScore,
        role: user.role,
        status: user.status,
        emailVerified: user.emailVerifiedAt !== null,
        termsAcceptedAt: user.termsAcceptedAt?.toISOString() ?? null,
        createdAt: user.createdAt.toISOString(),
      },
      reviews: reviews.map((review) => ({
        id: review.id,
        gameId: review.gameId,
        gameSlug: review.game.slug,
        gameName: review.game.name,
        recommended: review.recommended,
        rating: review.rating,
        text: review.text,
        hoursPlayed: review.hoursPlayed,
        createdAt: review.createdAt.toISOString(),
        updatedAt: review.updatedAt.toISOString(),
        status: review.status,
      })),
      votes: votes.map((vote) => ({
        reviewId: vote.reviewId,
        useful: vote.useful,
        createdAt: vote.createdAt.toISOString(),
      })),
      reports: reports.map((report) => ({
        id: report.id,
        reviewId: report.reviewId,
        reason: report.reason,
        details: report.details,
        status: report.status,
        createdAt: report.createdAt.toISOString(),
      })),
      reputationEvents: reputationEvents.map((event) => ({
        id: event.id,
        reason: event.reason,
        delta: event.delta,
        createdAt: event.createdAt.toISOString(),
      })),
    };
  }

  async deleteMe(userId: string, password: string): Promise<void> {
    const user = await this.users.findCredentialsById(userId);
    if (!user) {
      throw new NotFoundError(ERROR_CODES.USER_NOT_FOUND, 'Account not found');
    }

    let passwordMatches = false;
    try {
      passwordMatches = await verify(user.passwordHash, password, ARGON);
    } catch {
      passwordMatches = false;
    }
    if (!passwordMatches) {
      throw new UnauthorizedError(ERROR_CODES.CURRENT_PASSWORD_INVALID, 'Current password is wrong');
    }

    const shortId = user.id.replaceAll('-', '').slice(0, 8);
    const passwordHash = await hash(randomBytes(32).toString('hex'), ARGON);
    const now = new Date();

    await this.prisma.$transaction(async (tx) => {
      await tx.user.update({
        where: { id: userId },
        data: {
          deletedAt: now,
          email: `deleted+${user.id}@invalid.local`,
          username: `deleted_${shortId}`,
          displayName: null,
          bio: null,
          avatarUrl: null,
          passwordHash,
          emailVerifiedAt: null,
        },
      });
      await tx.refreshToken.updateMany({
        where: { userId, revokedAt: null },
        data: { revokedAt: now },
      });
      await tx.passwordResetToken.updateMany({
        where: { userId, usedAt: null },
        data: { usedAt: now },
      });
      await tx.emailVerificationToken.updateMany({
        where: { userId, usedAt: null },
        data: { usedAt: now },
      });
    });
  }

  private normaliseAvatarUrl(value: string | null): string | null {
    const trimmed = value?.trim() ?? '';
    if (!trimmed) return null;
    try {
      const parsed = new URL(trimmed);
      if (parsed.protocol !== 'https:') {
        throw new BadRequestError(ERROR_CODES.INVALID_AVATAR_URL, 'Avatar URL must be https');
      }
      return parsed.toString();
    } catch (error) {
      if (error instanceof BadRequestError) throw error;
      throw new BadRequestError(ERROR_CODES.INVALID_AVATAR_URL, 'Avatar URL is not valid');
    }
  }
}

function emptyToNull(value: string | null): string | null {
  const trimmed = value?.trim() ?? '';
  return trimmed.length === 0 ? null : trimmed;
}

File: apps/api/src/GameScore.Infrastructure/Users/UsersService.cs
using GameScore.Application.Errors;
using GameScore.Application.Http;
using GameScore.Domain.Enums;
using GameScore.Infrastructure.Auth;
using GameScore.Infrastructure.Integrations;
using GameScore.Infrastructure.Mappers;
using GameScore.Infrastructure.Persistence;
using GameScore.Infrastructure.Reviews;
using Microsoft.EntityFrameworkCore;
using System.Text.Json;

namespace GameScore.Infrastructure.Users;

public sealed class UsersService
{
    private readonly GameScoreDbContext _db;
    private readonly ReviewDataAccess _reviews;
    private readonly PasswordHasher _passwords;

    public UsersService(GameScoreDbContext db, ReviewDataAccess reviews, PasswordHasher passwords)
    {
        _db = db;
        _reviews = reviews;
        _passwords = passwords;
    }

    public async Task<object> GetProfileAsync(string username, CancellationToken cancellationToken = default)
    {
        var user = await FindPublicUserAsync(username, cancellationToken);
        var published = _db.Reviews.AsNoTracking()
            .Where(r => r.UserId == user.Id && r.DeletedAt == null && r.Status == DbReviewStatus.PUBLISHED);

        var agg = await published
            .GroupBy(r => r.Recommended)
            .Select(g => new { Recommended = g.Key, Count = g.Count() })
            .ToListAsync(cancellationToken);

        var recommended = agg.Where(a => a.Recommended).Sum(a => a.Count);
        var notRecommended = agg.Where(a => !a.Recommended).Sum(a => a.Count);
        var totalReviews = recommended + notRecommended;

        var usefulVotes = await published.SumAsync(r => r.UsefulCount, cancellationToken);
        var hours = await published.Where(r => r.HoursPlayed != null).SumAsync(r => r.HoursPlayed ?? 0, cancellationToken);
        var gamesReviewed = await published.Select(r => r.GameId).Distinct().CountAsync(cancellationToken);

        return new
        {
            id = user.Id,
            username = user.Username,
            displayName = user.DisplayName,
            avatarUrl = user.AvatarUrl,
            reputationScore = user.ReputationScore,
            role = user.Role.ToString(),
            deleted = false,
            bio = user.Bio,
            createdAt = user.CreatedAt.ToString("O"),
            stats = new
            {
                totalReviews,
                recommendedCount = recommended,
                notRecommendedCount = notRecommended,
                recommendationPercentage = totalReviews == 0 ? 0 : Math.Round(recommended / (double)totalReviews * 10000) / 100,
                usefulVotesReceived = usefulVotes,
                gamesReviewed,
                totalHoursPlayed = hours,
            },
        };
    }

    public async Task<object> ListReviewsAsync(
        string username,
        string? cursor,
        int? limit,
        Guid? viewerId,
        CancellationToken cancellationToken = default)
    {
        var user = await FindPublicUserAsync(username, cancellationToken);
        var pageSize = Pagination.ClampLimit(limit, Pagination.DefaultCursorSize);
        var (items, nextCursor) = await _reviews.ListByUserAsync(user.Id, cursor, pageSize, cancellationToken);
        var mapped = await MapReviewsAsync(items, viewerId, cancellationToken);
        return new { items = mapped, meta = Pagination.CursorMeta(pageSize, nextCursor) };
    }

    public async Task<object> ExportMeAsync(Guid userId, CancellationToken cancellationToken = default)
    {
        var user = await _db.Users.AsNoTracking()
            .FirstOrDefaultAsync(u => u.Id == userId && u.DeletedAt == null, cancellationToken);
        if (user is null)
        {
            throw new NotFoundException(ErrorCodes.UserNotFound, "Account not found");
        }

        var reviews = await _db.Reviews.AsNoTracking()
            .Include(r => r.Game)
            .Where(r => r.UserId == userId)
            .OrderByDescending(r => r.CreatedAt)
            .ToListAsync(cancellationToken);

        var votes = await _db.ReviewVotes.AsNoTracking()
            .Where(v => v.UserId == userId)
            .OrderByDescending(v => v.CreatedAt)
            .ToListAsync(cancellationToken);

        var reports = await _db.ReviewReports.AsNoTracking()
            .Where(r => r.ReporterId == userId)
            .OrderByDescending(r => r.CreatedAt)
            .ToListAsync(cancellationToken);

        var reputationEvents = await _db.ReputationEvents.AsNoTracking()
            .Where(e => e.UserId == userId)
            .OrderByDescending(e => e.CreatedAt)
            .ToListAsync(cancellationToken);

        return new
        {
            exportedAt = DateTime.UtcNow.ToString("O"),
            account = new
            {
                email = user.Email,
                username = user.Username,
                displayName = user.DisplayName,
                bio = user.Bio,
                avatarUrl = user.AvatarUrl,
                reputationScore = user.ReputationScore,
                role = user.Role.ToString(),
                status = user.Status.ToString(),
                emailVerified = user.EmailVerifiedAt is not null,
                termsAcceptedAt = user.TermsAcceptedAt?.ToString("O"),
                createdAt = user.CreatedAt.ToString("O"),
            },
            reviews = reviews.Select(r => new
            {
                id = r.Id,
                gameId = r.GameId,
                gameSlug = r.Game.Slug,
                gameName = r.Game.Name,
                recommended = r.Recommended,
                rating = r.Rating,
                text = r.Text,
                hoursPlayed = r.HoursPlayed,
                createdAt = r.CreatedAt.ToString("O"),
                updatedAt = r.UpdatedAt.ToString("O"),
                status = r.Status.ToString(),
            }),
            votes = votes.Select(v => new
            {
                reviewId = v.ReviewId,
                useful = v.Useful,
                createdAt = v.CreatedAt.ToString("O"),
            }),
            reports = reports.Select(r => new
            {
                id = r.Id,
                reviewId = r.ReviewId,
                reason = r.Reason.ToString(),
                details = r.Details,
                status = r.Status.ToString(),
                createdAt = r.CreatedAt.ToString("O"),
            }),
            reputationEvents = reputationEvents.Select(e => new
            {
                id = e.Id,
                reason = e.Reason.ToString(),
                delta = e.Delta,
                balanceAfter = e.BalanceAfter,
                createdAt = e.CreatedAt.ToString("O"),
            }),
        };
    }

    public async Task<object> UpdateMeAsync(
        Guid userId,
        string? displayName,
        string? bio,
        string? avatarUrl,
        CancellationToken cancellationToken = default)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == userId, cancellationToken);
        if (user is null)
        {
            throw new NotFoundException(ErrorCodes.UserNotFound, "Account not found");
        }

        if (displayName is not null)
        {
            user.DisplayName = string.IsNullOrWhiteSpace(displayName) ? null : displayName.Trim();
        }

        if (bio is not null)
        {
            user.Bio = string.IsNullOrWhiteSpace(bio) ? null : bio.Trim();
        }

        if (avatarUrl is not null)
        {
            if (!string.IsNullOrWhiteSpace(avatarUrl) && !CdnImageRules.IsAllowedImageUrl(avatarUrl))
            {
                throw new BadRequestException(ErrorCodes.InvalidAvatarUrl, "Avatar URL must be HTTPS from an allowed host");
            }

            user.AvatarUrl = string.IsNullOrWhiteSpace(avatarUrl) ? null : avatarUrl.Trim();
        }

        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        return Mappers.UserMapper.ToAuthenticatedUser(user);
    }

    public async Task DeleteMeAsync(Guid userId, string password, CancellationToken cancellationToken = default)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == userId, cancellationToken);
        if (user is null)
        {
            return;
        }

        if (!await _passwords.VerifyAsync(user.PasswordHash, password, cancellationToken))
        {
            throw new UnauthorizedException(ErrorCodes.InvalidCredentials, "Password is incorrect");
        }

        user.DeletedAt = DateTime.UtcNow;
        user.Email = $"deleted-{user.Id:N}@anonymous.local";
        user.Username = $"deleted-{user.Id:N}"[..Math.Min(32, $"deleted-{user.Id:N}".Length)];
        user.DisplayName = null;
        user.Bio = null;
        user.AvatarUrl = null;
        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
    }

    private async Task<UserEntity> FindPublicUserAsync(string username, CancellationToken cancellationToken)
    {
        var user = await _db.Users.AsNoTracking()
            .FirstOrDefaultAsync(u => u.Username == username.Trim().ToLowerInvariant(), cancellationToken);
        if (user is null || user.DeletedAt is not null)
        {
            throw new NotFoundException(ErrorCodes.UserNotFound, "User not found");
        }

        return user;
    }

    private async Task<IReadOnlyList<object>> MapReviewsAsync(
        IReadOnlyList<ReviewBundle> items,
        Guid? viewerId,
        CancellationToken cancellationToken)
    {
        var ids = items.Select(i => i.Review.Id).ToList();
        var context = viewerId is null
            ? (Votes: new Dictionary<Guid, bool>(), Reports: new HashSet<Guid>())
            : await _reviews.ViewerContextAsync(ids, viewerId.Value, cancellationToken);

        return items.Select(item =>
        {
            bool? viewerVote = null;
            if (viewerId is not null && context.Votes.TryGetValue(item.Review.Id, out var vote))
            {
                viewerVote = vote;
            }

            return ReviewMapper.ToDto(
                item.Review,
                item.Author,
                item.Game,
                item.Platform,
                viewerId,
                viewerVote,
                context.Reports.Contains(item.Review.Id));
        }).ToList();
    }
}
